rfcs/2021-06-01-7694-improve-monitor-build-performance.md

```diff
@@ -0,0 +1,170 @@
+users building Rust projects today, they're using their system linker, which is single-threaded.
+However, newer linkers exist that can exploit the parallelism of today's multi-core systems and
+provide much faster link times.
+
+We should explore the use of [lld](https://lld.llvm.org/), which is a linker from the LLVM project
+itself, the compiler toolchain that underpins Rust itself. `lld` is optimized for multi-core
+systems, and generally more performance-focused than system linkers like GNU's `gold` linker.  In
+some cases, `lld` shows a 3-5x improvement over `gold`, which represents a potential link time
+reduction on the order of _minutes_ when used for Vector.
+
+While Rust is moving towards bundling and using LLD
+[directly](https://github.com/rust-lang/rust/issues/39915), and by default, that work is not yet
+complete.  However, it is trivial to manually specify using `lld` directly when building a Rust
+project, and usage by developers building Rust projects already shows that it is fairly stable
+and is able to successfully link many large projects.
+
+### Rework “release” profile for non-versioned releases
+
+We currently utilize the same Cargo “release” profile whether we’re minting a versioned release or
+simply running a release build locally for performance testing.  While a boon for our users, our
+usage of LTO and modified “codegen-units” represents a significant increase in compilation time.  On
+a standard-issue laptop used by Vector engineers, it can take upwards of 40 minutes to build a
+release binary with these settings enabled.  With the settings disabled, the same build takes around
+27 minutes, or 35% less time.
+
+These settings can be added back during the build process, when building versioned releases, without
+much effort.
+
+### Add system telemetry to Github Actions runners
+
+As Github Actions natively provides no telemetry of any kind about runners, or even basic metrics
+(how long did a job sit before running? etc), we’re often in the dark when it comes to CI and runner
+performance as a whole.
+
+We should be running the Datadog agent on every runner possible, collecting both high-level system
+metrics as well as more fine-grained metrics, such as Docker metrics on a per-container basis.
+These metrics will, I believe, prove invaluable to debugging issues with the overall CI pipeline and
+its performance over time.  While understanding the build performance will likely come down to more
+advanced tooling used directly by engineers, we can’t optimize things like “use all the cores
+available” or “we’re using too much memory during CI runs” unless we actually have some data, any
+data, from those runs.
+
+### Manually instrument CI to report build performance
+
+Once we were properly utilizing Datadog on our CI runners, we could then instrument CI builds
+directly to report the build time to Datadog, allowing us to track the build performance over time.
```

Comment:
This will probably be a really big deal over time, as it will surface both long term trends and discontinuities that need attention.